scylla-cql/src/frame/response/result.rs

```diff
@@ -1187,28 +1188,32 @@ impl RawMetadataAndRawRows {
     fn metadata_deserializer(
         col_count: usize,
         global_tables_spec: bool,
-    ) -> impl for<'frame> FnOnce(&mut &'frame [u8]) -> StdResult<ResultMetadata<'frame>, RowsParseError>
-    {
+    ) -> impl for<'frame> FnOnce(
+        &mut &'frame [u8],
+    ) -> StdResult<
+        ResultMetadata<'frame>,
+        ResultMetadataLazyDeserializationError,
+    > {
```

Comment:
I don't understand why we need another error type here. Why not return ResultMetadataParseError? Why do we even need ResultMetadataLazyDeserializationError instead of just using ResultMetadataParseError?

Reply:
`ResultMetadataParseError` is a broader type that is used to parse result prepared metadata eagerly (after PREPARE request).

Reply:
It is used to parse result metadata, and here we also parse result metadata, just in a different way. Error structure doesn't have to reflect our internal code structure - I'd say its undesirable.

Reply:
So do you suggest to reuse `ResultMetadataParseError` and return a broader error type in such case?

Reply:
I'm not insisting on it, but I don't really see the point of introducing second error for basically the same operation. @wprzytula ?

Reply:
Notice that `ResultMetadataParseError` is not really a superset of new error. In lazy deserialization we also try to deserialize rows count, which is not applicable for prepared statement's result metadata (in response to PREPARE request).

Reply:
And we also have a column count check in new error:
```rust
            if server_metadata.col_count() != server_metadata.col_specs().len() {
                return Err(
                    ResultMetadataLazyDeserializationError::ColumnCountMismatch {
                        col_count: server_metadata.col_count(),
                        col_specs_count: server_metadata.col_specs().len(),
                    },
                );
            }
```

But this could apply to the result metadata in `PREPARED` response as well. Not sure why we don't perform the same check there.

Reply:
The better question is why do we have this check here. 
From protocol spec:
```
    - <metadata> is composed of:
        <flags><columns_count>[<paging_state>][<global_table_spec>?<col_spec_1>...<col_spec_n>]
```

deser_col_specs_generic takes col_count  and deserializes column specs that many times. The only possibility for this check to be triggered is a bug in the driver - not invalid bytes from the network.

Reply:
I don't see the point of having user-facing error type that can only indicate a serious bug in the driver.

Reply:
True.

Reply:
Before refactor (0.13) it looked like this: https://github.com/scylladb/scylla-rust-driver/blob/v0.13.0/scylla-cql/src/frame/response/result.rs#L927-L940

Reply:
And I think your claim is true for previous version as well. `deser_col_specs` would take a `column_count` and  deserialize col specs `column_count` times: https://github.com/scylladb/scylla-rust-driver/blob/v0.13.0/scylla-cql/src/frame/response/result.rs#L554-L575